Add documentation for VolumeSnapshot Beta (#17233)
kubernetes/website: content/en/docs/concepts/storage/volume-snapshots.md

---
reviewers:
- saad-ali
- thockin
- msau42
- jingxu97
- xing-yang
- yuxiangqian
title: Volume Snapshots
content_template: templates/concept
weight: 20
---

{{% capture overview %}}

> 

Reviewer: ```suggestion
{{< feature-state for_k8s_version="1.17" state="beta" >}}
```

> This document describes the current state of `VolumeSnapshots` in Kubernetes. Familiarity with [persistent volumes](/docs/concepts/storage/persistent-volumes/) is suggested. `VolumeSnapshot` is introduced as an alpha feature in Kubernetes 1.12 and is promoted to beta in 1.17.

Reviewer: If you use a feature state shortcode, this could be:
```suggestion
In Kubernetes, a _VolumeSnapshot_ represents a snapshot of a storage volume. This document assumes that you are already familiar with Kubernetes [persistent volumes](/docs/concepts/storage/persistent-volumes/).
```

---
reviewers:
- saad-ali
- thockin
- msau42
- jingxu97
- xing-yang
- yuxiangqian
title: Volume Snapshots
content_template: templates/concept
weight: 20
---

{{% capture overview %}}

{{< feature-state for_k8s_version="1.17" state="beta" >}}

> In Kubernetes, a _VolumeSnapshot_ represents a snapshot of a storage volume. This document assumes that you are already familiar with Kubernetes [persistent volumes](/docs/concepts/storage/persistent-volumes/).

Reviewer 2: snapshot of a PersistentVolumeClaim?

Author: Changed to "snapshot of a volume on a storage system"

{{% capture overview %}}

> {{< feature-state for_k8s_version="v1.12" state="alpha" >}}

Reviewer: I don't think this line should remain/

Author: Removed

A `VolumeSnapshot` is a request for snapshot of a volume by a user. It is similar to a PersistentVolumeClaim.

> `VolumeSnapshotClass` allows user to specify different attributes belonging to a `VolumeSnapshot`. These attibutes may differ among snapshots taken from the same volume on the storage system and therefore cannot be expressed by using the same `StorageClass` of a `PersistentVolumeClaim`.

Reviewer: nit:
```suggestion
`VolumeSnapshotClass` allows you to specify different attributes belonging to a `VolumeSnapshot`. These attibutes may differ among snapshots taken from the same volume on the storage system and therefore cannot be expressed by using the same `StorageClass` of a `PersistentVolumeClaim`.
```

Author: Done

* CSI drivers may or may not have implemented the volume snapshot functionality. The CSI drivers that have provided support for volume snapshot will likely use the csi-snapshotter. See [CSI Driver documentation](https://kubernetes-csi.github.io/docs/) for details.
* The CRDs and snapshot controller installations are the responsibility of the Kubernetes distribution.

## Lifecycle of a volume snapshot and volume snapshot content

`VolumeSnapshotContents` are resources in the cluster. `VolumeSnapshots` are requests for those resources. The interaction between `VolumeSnapshotContents` and `VolumeSnapshots` follow this lifecycle:

### Provisioning Volume Snapshot

There are two ways snapshots may be provisioned: pre-provisioned or dynamically provisioned.

#### Pre-provisioned {#static}
A cluster administrator creates a number of `VolumeSnapshotContents`. They carry the details of the real volume snapshot on the storage system which is available for use by cluster users. They exist in the Kubernetes API and are available for consumption.

#### Dynamic

> Instead of a pre-existing snapshot, a user can request that a snapshot of a PVC is dynamically taken. The [VolumeSnapshotClass](/docs/concepts/storage/volume-snapshot-classes/) specifies storage provider-specific parameters to use when taking a snapshot.

Reviewer: Is this wording better:
```suggestion
Instead of using a pre-existing snapshot, you can request that the cluster creates a snapshot of an existing PersistentVolumeClaim. The [VolumeSnapshotClass](/docs/concepts/storage/volume-snapshot-classes/) specifies storage provider-specific parameters to use when taking a snapshot.
```
? (I'm not sure - although, if I've misunderstood, then that's a hint that the current wording is not clear).

Author: Reworded.

### Binding

The snapshot controller handles the binding of a `VolumeSnapshot` object with an appropriate `VolumeSnapshotContent` object, in both pre-provisioned and dynamically provisioned scenarios. The binding is a one-to-one mapping.

VolumeSnapshots will remain unbound indefinitely if a matching VolumeSnapshotContent does not exist. VolumeSnapshots will be bound as matching VolumeSnapshotContents become available.

### Persistent Volume Claim as Snapshot Source Protection

> The purpose of the Persistent Volume Claim Object while a snapshot is being taken Protection feature is to ensure that in-use PVC API objects are not removed from the system (as this may result in data loss).

Reviewer: > The purpose of the Persistent Volume Claim Object while a snapshot is being taken Protection feature is to ensure that in-use PVC API objects are not removed from the system (as this may result in data loss).
Ouch, that's hard to parse.

How about:
```suggestion
Kubernetes protects PersistentVolumeClaim objects during snapshotting operations.
```

(if that protection isn't done using a built-in controller, then replace “Kubernetes” with a description of what implements the protection).

Author: I shortened the sentence a little bit.  Hope that reads better.

> If a snapshot is being taken of a PVC, that PVC is in-use. If a user deletes a PVC API object in active use as a snapshot source, the PVC object is not removed immediately. Instead, removal of the PVC object is postponed until the snapshot is readyToUse or aborted.

Reviewer: ```suggestion
While a snapshot is being taken of a PVC, that PVC is in-use. If you delete a PersistentVolumeClaim that is in active use as a snapshot source, the PersistentVolumeClaim is not removed immediately. Instead, removal of the PVC object is postponed until the snapshot either becomes readyToUse or is aborted.
```
How about this?

Author: Done

### Delete

Deletion is triggered by deleting the `VolumeSnapshot` object, and the `DeletionPolicy` will be followed. If the `DeletionPolicy` is `Delete`, then the underlying storage snapshot will be deleted along with the `VolumeSnapshotContent` object. If the `DeletionPolicy` is `Retain`, then both the underlying snapshot and `VolumeSnapshotContent` remain.

## Volume Snapshot Contents

> Each VolumeSnapshotContent contains a spec and status. In dynamic provisioning, the `VolumeSnapshotContent` object is created by the snapshot common controller as follows.

Reviewer: Does this work better:
```suggestion
Each VolumeSnapshotContent contains a spec and status. In dynamic provisioning, the snapshot common controller creates `VolumeSnapshotContent` objects. Here's an example:
```
?

Author: Done

> Users need to be aware of the following when using this feature:

Reviewer 2: The previous 3 paragraphs may need to be reworded to fit the new model, especially the 3rd paragraph "While `VolumeSnapshots` allow a user to consume abstract storage resources..."

Author: Reworded.

* API Objects `VolumeSnapshot`, `VolumeSnapshotContent`, and `VolumeSnapshotClass` are {{< glossary_tooltip term_id="CustomResourceDefinition" text="CRDs" >}}, not part of the core API.
* `VolumeSnapshot` support is only available for CSI drivers.
* As part of the deployment process in the beta version of `VolumeSnapshot`, the Kubernetes team provides a snapshot controller to be deployed into the control plane, and a sidecar helper container called csi-snapshotter to be deployed together with the CSI driver.  The snapshot controller watches `VolumeSnapshot` and `VolumeSnapshotContent` objects and is responsible for the creation and deletion of `VolumeSnapshotContent` object in dynamic provisioning.  The sidecar csi-snapshotter watches `VolumeSnapshotContent` objects and triggers `CreateSnapshot` and `DeleteSnapshot` operations against a CSI endpoint.

> * CSI drivers may or may not have implemented the volume snapshot functionality. The CSI drivers that have provided support for volume snapshot will likely use the snapshot controller and the csi-snapshotter.

Reviewer 2: CSI drivers should only use the sidecar.  Link to the sidecar repo/documentation?

Author: Done

> * You need to manually install the `VolumeSnapshot` v1beta CRDs.

Reviewer 2: "The CRDs and snapshot controller installations are the responsibility of the Kubernetes distribution, not the CSI drivers."

Author: Sure.  Should we add "check its availability with the specific Kubernetes distribution you are using"?  Otherwise it may not be there?

## Lifecycle of a volume snapshot and volume snapshot content

`VolumeSnapshotContents` are resources in the cluster. `VolumeSnapshots` are requests for those resources. The interaction between `VolumeSnapshotContents` and `VolumeSnapshots` follow this lifecycle:

### Provisioning Volume Snapshot

There are two ways snapshots may be provisioned: pre-provisioned or dynamically provisioned.

> #### Pre-provisioned {#static}

Reviewer 2: remove static

Author: There's a previous comments asking me to add this link to static here so that any old links would still work:)

Reviewer: If it's better to break existing links in favour of clear naming, as a conscious decision, that's OK by me.

> A cluster administrator creates a number of `VolumeSnapshotContents`. They carry the details of the real storage which is available for use by cluster users. They exist in the Kubernetes API and are available for consumption.

Reviewer 2: "real storage" => "real storage snapshots"?

Author: Done

#### Dynamic

> The dynamic provisioning is based on `VolumeSnapshotClasses`: the `VolumeSnapshot` must request a

Reviewer 2: Instead of a pre-existing snapshot, a user can request that a snapshot of a PVC is dynamically taken. The `VolumeSnapshotClass` specifies storage provider-specific parameters to use when taking a snapshot.

Author: Done

[volume snapshot class](/docs/concepts/storage/volume-snapshot-classes/) and
the administrator must have created and configured that class in order for dynamic
provisioning to occur.

### Binding

> A user creates, or has already created in the case of dynamic provisioning, a `VolumeSnapshot` with a specific attributes defined by the `VolumeSnapshotClass`. A control loop watches for new VolumeSnapshots, finds a matching VolumeSnapshotContent (if possible), and binds them together. If a VolumeSnapshotContent was dynamically provisioned for a new VolumeSnapshot, the loop will always bind that VolumeSnapshotContent to the VolumeSnapshot. Once bound, `VolumeSnapshot` binds are exclusive, regardless of how they were bound. A VolumeSnapshot to VolumeSnapshotContent binding is a one-to-one mapping.

Reviewer 2: I would remove a lot of the implementation detail from this section and reword it something like 

"The snapshot controller handles binding a `VolumeSnapshot` object with an appropriate `VolumeSnapshotContent' object, in both pre-provisioned and dynamically provisioned scenarios. The binding is a one-to-one mapping."

Author: Done

> VolumeSnapshots will remain unbound indefinitely if a matching VolumeSnapshotContent does not exist. VolumeSnapshots will be bound as matching VolumeSnapshotContents become available.

Reviewer 2: "In the pre-provisioned case, "

### Persistent Volume Claim as Snapshot Source Protection

> The purpose of the Persistent Volume Claim Object as Snapshot Source Protection feature is to ensure that in-use PVC API objects are not removed from the system (as this may result in data loss).

Reviewer 2: while a snapshot is being taken

Author: Done

> If a PVC is in active use by a snapshot as a source to create the snapshot, the PVC is in-use. If a user deletes a PVC API object in active use as a snapshot source, the PVC object is not removed immediately. Instead, removal of the PVC object is postponed until the PVC is no longer actively used by any snapshots. A PVC is no longer used as a snapshot source when `ReadyToUse` of the snapshot `Status` becomes `true`.

Reviewer 2: "If a snapshot is being taken of a PVC, that PVC is in use"

"Instead, removal of the PVC object is postponed until the snapshot is ready or aborted"

Author: Done

> 

Reviewer 2: For deletion section below,

"Deletion is triggered by deleting the VolumeSnapshot object, and the DeletionPolicy will be followed. If the DeletionPolicy is 'Delete', then the underlying storage snapshot will be deleted along with the VolumeSnapshotContent object. If the DeletionPolicy is 'Retain', then both the underlying snapshot and VolumeSnapshotContent remain"

Author: Done

* API Objects `VolumeSnapshot`, `VolumeSnapshotContent`, and `VolumeSnapshotClass` are CRDs, not part of the core API.
* `VolumeSnapshot` support is only available for CSI drivers.

> * As part of the deployment process in `VolumeSnapshot` API v1beta1, the Kubernetes team provides two external snapshot controllers called `external-snapshotter`, a common controller to be deployed on the controller node and a sidecar helper container to be deployed together with the CSI driver.  The common controller watches `VolumeSnapshot` and `VolumeSnapshotContent` objects and is responsible for the creation and deletion of `VolumeSnapshotContent` object in dynamic provisioning.  The sidecar controller watches `VolumeSnapshotContent` objects and triggers `CreateSnapshot` and `DeleteSnapshot` operations against a CSI endpoint.

Reviewer: Optional suggestion (not particularly relevant to v1.17, but a good idea): add a glossary tooltip for “controller”:
```suggestion
* As part of the deployment process in `VolumeSnapshot` API v1beta1, the Kubernetes team provides two external snapshot {{< glossary_tooltip term_id="controller" text="controllesr" >}} called `external-snapshotter`, a common controller to be deployed into the control plane, and a sidecar helper container to be deployed together with the CSI driver.  The common controller watches `VolumeSnapshot` and `VolumeSnapshotContent` objects and is responsible for the creation and deletion of `VolumeSnapshotContent` object in dynamic provisioning.  The sidecar controller watches `VolumeSnapshotContent` objects and triggers `CreateSnapshot` and `DeleteSnapshot` operations against a CSI endpoint.
```

Reviewer: I'm not sure I understand the “_in `VolumeSnapshot` API v1beta1_” addition. What's the learning outcome for a reader seeing that phrase?

Reviewer: Aside: I've got an open PR (#16885) to add reference documentation for controllers. Would it make sense to put the implementation details into the reference section once that's merged?

Author: The snapshot controller code repo is here: https://github.com/kubernetes-csi/external-snapshotter, not in the core kubernetes repo.  Do you want to include out-of-tree controllers in the reference doc?

On the other hand, since we documented the snapshot controller here, it does make sense to reference it in your PR.

Reviewer: I think it could be a separate PR and a section called something like “Additional controllers”, to highlight that this doesn't run inside `kube-controller-manager`.

* CSI drivers may or may not have implemented the volume snapshot functionality. The CSI drivers that have provided support for volume snapshot will likely use `external-snapshotter`.
* The `VolumeSnapshot` v1beta CRDs need to be installed manually by the CSI drivers.

## Lifecycle of a volume snapshot and volume snapshot content

`VolumeSnapshotContents` are resources in the cluster. `VolumeSnapshots` are requests for those resources. The interaction between `VolumeSnapshotContents` and `VolumeSnapshots` follow this lifecycle:

### Provisioning Volume Snapshot

There are two ways snapshots may be provisioned: pre-provisioned or dynamically provisioned.

> #### Pre-provisioned

Reviewer: Optional tweak:
```suggestion
#### Pre-provisioned {#static}
```
(if there are existing inbound hyperlinks, this change means those existing links will stay working).

> * API Objects `VolumeSnapshot`, `VolumeSnapshotContent`, and `VolumeSnapshotClass` are CRDs, not part of the core API.

Reviewer: ```suggestion
 * API Objects `VolumeSnapshot`, `VolumeSnapshotContent`, and `VolumeSnapshotClass` are {{< glossary_tooltip term_id="CustomResourceDefinition" text="CRDs" >}}, not part of the core API.
```

Author: Done

* `VolumeSnapshot` support is only available for CSI drivers.
* As part of the deployment process in the beta version of `VolumeSnapshot`, the Kubernetes team provides a snapshot controller to be deployed into the control plane, and a sidecar helper container called csi-snapshotter to be deployed together with the CSI driver.  The snapshot controller watches `VolumeSnapshot` and `VolumeSnapshotContent` objects and is responsible for the creation and deletion of `VolumeSnapshotContent` object in dynamic provisioning.  The sidecar csi-snapshotter watches `VolumeSnapshotContent` objects and triggers `CreateSnapshot` and `DeleteSnapshot` operations against a CSI endpoint.
* CSI drivers may or may not have implemented the volume snapshot functionality. The CSI drivers that have provided support for volume snapshot will likely use the snapshot controller and the csi-snapshotter.

> * The `VolumeSnapshot` v1beta CRDs need to be installed manually.

Reviewer: ```suggestion
* You need to manually install the `VolumeSnapshot` v1beta CRDs.
```
(this kind of relies on the glossary tooltip for CRD being added above)

Author: Done

## Volume Snapshot Contents

> Each VolumeSnapshotContent contains a spec and status, which is the specification and status of the volume snapshot content. In dynamic provisioning, the `VolumeSnapshotContent` object is created by the snapshot common controller.

Reviewer 2: 2nd half of this sentence is sort of redundant...

"In the pre-provisioned snapshot case, the object is created by an administrator"

Author: Done

```yaml
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotContent
metadata:
  name: snapcontent-72d9a349-aacd-42d2-a240-d775650d2455
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:

>     volumeHandle: ee0cfb94-f8d4-11e9-b2d8-0242ac110002

Reviewer 2: Are these fields going to be explained?

Author: Added explanation.

  volumeSnapshotClassName: csi-hostpath-snapclass
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
    uid: 72d9a349-aacd-42d2-a240-d775650d2455
```

For pre-provisioned snapshots, you (as cluster administrator) are responsible for creating the `VolumeSnapshotContent` object.

```yaml
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotContent
metadata:
  name: new-snapshot-content-test
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

## VolumeSnapshots

> Each VolumeSnapshot contains a spec and a status, which is the specification and status of the volume snapshot.

Reviewer 2: 2nd half of this sentence is redundant

Author: Done

```yaml
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshot
metadata:
  name: new-snapshot-test
spec:
  volumeSnapshotClassName: csi-hostpath-snapclass
  source:

>     persistentVolumeClaimName: pvc-test

Reviewer 2: Maybe it's worth splitting up this section into "How to request a dynamically provisioned snapshot" and "How to request a pre-existing snapshot", and give examples for both.

Author: I'd like to submit another PR as a task page for that.  Things described at Concepts page seem to be relatively high level.

Reviewer: :+1: for a separate task page.

Reviewer 2: I guess it's a little strange that not all the fields in the VolumeSnapshot are not explained, considering that this is the user-facing object. It feels inconsistent with the VolumeSnapshotContent object above that is very detailed but actually not really that important to the user.

```

### Class

A volume snapshot can request a particular class by specifying the name of a
[VolumeSnapshotClass](/docs/concepts/storage/volume-snapshot-classes/)

> using the attribute `volumeSnapshotClassName`.

Reviewer 2: If nothing is set, then the default class is used.

Author: Done

## Volume Snapshot Contents

Each VolumeSnapshotContent contains a spec and status, which is the specification and status of the volume snapshot content. In dynamic provisioning, the `VolumeSnapshotContent` object is created by the snapshot common controller.

```yaml
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotContent
metadata:
  name: snapcontent-72d9a349-aacd-42d2-a240-d775650d2455
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    volumeHandle: ee0cfb94-f8d4-11e9-b2d8-0242ac110002
  volumeSnapshotClassName: csi-hostpath-snapclass
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
    uid: 72d9a349-aacd-42d2-a240-d775650d2455
```

> For pre-provisioned snapshots, the admin is responsible for creating the `VolumeSnapshotContent` object.

Reviewer: ```suggestion
For pre-provisioned snapshots, you (as cluster administrator) are responsible for creating the `VolumeSnapshotContent` object.
```
per [address the reader as “you”](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you).

Author: Done
